server/pulp/plugins/file/distributor.py

```diff
@@ -173,11 +174,27 @@ def publish_repo_fast_forward(self, repo, publish_conduit, config):
                             unit_checksum_old_set.add(checksum)
                             if checksum not in unit_checksum_set:
                                 unit_over_path_map[checksum] = fields[0]
+                _logger.debug("%d items were in MANIFEST %s, which exists? %s." % (
+                              len(unit_checksum_old_set), metadata_filename,
+                              os.path.exists(metadata_filename)))
 
                 # Copy incremental files into publishing directories
                 checksum_absent_set = unit_checksum_set - unit_checksum_old_set
+                _logger.debug("Increasing %d units" % len(checksum_absent_set))
+
+                # If added too many units, then publish repo with force_full
+                max_increase_units = min(50000, len(units) / len(hosting_locations))
```

Comment:
What's the idea with dividing by len(hosting_locations)?

Reply:
During publish, symlink generating and copying to hosting_locations take most of the time. so both are expected to do less.

For symlink generating, when doing with fast-forward,
`total number of generated symlinks` = `number of increase units` * `len(hosting_locations)`
When doing with force full,
`total number of generated symlinks` = `number of total units`

We expect to do publish with fast-forward to gain better performance, so,
`number of increase units` * `len(hosting_locations)` <= `number of total units`
Then, 
`number of increase units` <= `number of total units / len(hosting_locations)`